### PR TITLE
provider/nscloud: exchange user tokens to tenant tokens for cluster ops

### PR DESCRIPTION
**Pull request by edganiukov**
Fixes NSL-500

### CODE REVIEW

File: internal/providers/nscloud/api/rpc.go [hunk 1]
```diff
@@ -34,22 +34,11 @@ type API struct {
...
-	if err != nil {
-		if os.IsNotExist(err) {
-			// Pass user auth for now.
-			// TODO exchange user auth for tenant tokens.
-			return auth.GenerateToken(ctx)
-		}
-
-		return "", err
+func MakeAPI(endpoint string) API {
+	fetchTenantToken := func(ctx context.Context) (string, error) {
```
Comment (n-g) [inline internal/providers/nscloud/api/rpc.go:38]: Let's inline this.

File: internal/providers/nscloud/api/rpc.go [hunk 2]
```diff
@@ -290,6 +285,7 @@ func ExchangeToken(ctx context.Context, scopes ...string) (string, error) {
 		return "", err
 	}
 
+	// No tenant token stored, so use user token and exchange it to a tenant token.
```
Comment (n-g) [inline internal/providers/nscloud/api/rpc.go:288]: Should we cache unscoped tenant tokens?

File: internal/providers/nscloud/api/rpc.go (hunk 1 above)
Comment (edganiukov) [reply internal/providers/nscloud/api/rpc.go:38]: You mean use `FetchToken: ExchangeToken` the `fnapi.Call` objects? It doesn't work due to type (`ExchangeToken` has optional `scopes` parameter).

Comment (n-g) [reply internal/providers/nscloud/api/rpc.go:38]: 🙈 yep - true - thanks!

File: internal/providers/nscloud/api/rpc.go [hunk 3]
```diff
@@ -276,8 +265,14 @@ func ListClusters(ctx context.Context, api API) (*KubernetesClusterList, error)
...
 func ExchangeToken(ctx context.Context, scopes ...string) (string, error) {
+	// Check if there is already tenant token stored.
 	tenantToken, err := auth.LoadTenantToken()
 	if err == nil {
+		// If no scopes provided we can immediately return the token.
+		if len(scopes) == 0 {
+			return tenantToken.TenantToken, nil
+		}
+
 		resp, err := fnapi.ExchangeTenantToken(ctx, tenantToken.TenantToken, scopes)
```
Comment (hugosantos) [inline internal/providers/nscloud/api/rpc.go:276]: Ideally calls that have non-trivial duration should be wrapped in tasks, so users can see where time is being spent.

File: internal/providers/nscloud/api/rpc.go (hunk 2 above)
Comment (edganiukov) [reply internal/providers/nscloud/api/rpc.go:288]: done

File: internal/auth/tokens.go [hunk 4]
```diff
@@ -57,5 +60,18 @@ func LoadTenantToken() (*Token, error) {
 		return nil, err
 	}
 
+	claims := jwt.RegisteredClaims{}
+	if _, _, err := new(jwt.Parser).ParseUnverified(token.TenantToken, claims); err != nil {
+		return nil, err
+	}
+
+	// If stored token is expired, we remove it and return [fs.ErrNotExist].
+	if !claims.VerifyExpiresAt(time.Now(), true) {
```
Comment (edganiukov) [inline internal/auth/tokens.go:69]: Wondering if I should have some leeway here?

Comment (nichtverstehen) [reply internal/auth/tokens.go:69]: You mean throw away a token that is still valid, but is about to expire? Would make sense.

But starting simple with a strict check is fine.

(Not the other way of course - we shouldn't try to use recently expired tokens).

Comment (edganiukov) [reply internal/auth/tokens.go:70]: > You mean throw away a token that is still valid, but is about to expire? Would make sense.
Yep, that is what I mean.

File: internal/providers/nscloud/api/rpc.go (hunk 3 above)
Comment (edganiukov) [reply internal/providers/nscloud/api/rpc.go]: done. I hope simply wrapping the whole func body is OK.

**Review by n-g — APPROVED** [review]

**Pull request merged**